### PR TITLE
Refs [TASK][41213] Handle event Home Screen

### PR DESCRIPTION
**Pull request by vuphong0x**
## Related Tickets
- [#Task 41213: Handle event Home Screen](https://edu-redmine.sun-asterisk.vn/issues/41213)
## WHAT
- Bind Data : get data pour into RecyclerView
- Click search bar: display Search Screen.
- Click item RecyclerView display List Meal Screen
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/efood/ui/detail/MealDetailFragment.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    override fun initEvents() {
+    }
+
+    override fun initData() {
+    }
+
+    companion object {
+        private const val MEAL = "MEAL"
+
```
Comment (trangnq-2533) [inline app/src/main/java/com/sun/efood/ui/detail/MealDetailFragment.kt]: xem lại convension về cách đặt tên bundle mục 2.10
https://github.com/framgia/coding-standards/blob/master/eng/android/codingstyleguide.md#210-string-constants-naming-and-values

File: app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt [hunk 2]
```diff
@@ -1,21 +1,107 @@
 package com.sun.efood.ui.home
 
+import android.util.Log
```
Comment (trangnq-2533) [inline app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt]: bỏ Log

File: app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt [hunk 3]
```diff
@@ -1,21 +1,107 @@
...
+import com.sun.efood.ui.search.SearchFragment
+import com.sun.efood.utils.addFragment
+import com.sun.efood.utils.gone
+import com.sun.efood.utils.show
 
-class HomeFragment : BaseFragment<FragmentHomeBinding>() {
+class HomeFragment : BaseFragment<FragmentHomeBinding>(), HomeContact.View {
+
+    private var presenter: HomePresenter? = null
+    private val randomMealAdapter = RandomMealAdapter(this::onRandomMealClick)
```
Comment (trangnq-2533) [inline app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt]:
```suggestion
    private val randomMealAdapter = RandomMealAdapter(::onRandomMealClick)
```

File: app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt [hunk 4]
```diff
@@ -1,21 +1,107 @@
...
+                    MealDaoImpl.getInstance(
+                        AppDatabase.getInstance(
+                            context
+                        )
+                    )
+                )
+            ),
+            MealCategoryRepository
+                .getInstance(MealCategoryRemoteDataSource.getInstance())
+        )
```
Comment (vietnh-1109) [inline app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt]: Cho riêng ra 1 file quản lí chuyên để khởi tạo repository

File: app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+            override fun onFailure(e: Exception?) {
+                view.showMessage(e?.message.toString())
+                view.hideLoading()
+            }
+        })
+    }
+
+    override fun getMealCategory() {
+        mealCategoryRepository.getMealCategories(object : OnDataCallback<List<MealCategory>> {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt]: show loading đâu

File: app/src/main/java/com/sun/efood/ui/home/adapter/MealCategoryAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.sun.efood.utils.loadImage
+
+class MealCategoryAdapter(private val onItemClick: (MealCategory) -> Unit) :
+    BaseAdapter<MealCategory, MealCategoryAdapter.ViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
+        ViewHolder(
+            ItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false),
+            onItemClick
+        )
```
Comment (vietnh-1109) [inline app/src/main/java/com/sun/efood/ui/home/adapter/MealCategoryAdapter.kt]:
```suggestion
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ViewHolder(
            ItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false),
            onItemClick
        )
```

File: app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt (hunk 5 above)
Comment (vuphong0x) [reply app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt]: Vì getRandomMeal() và getRandomMeal() đều sử dụng 1 progress bar nên em chỉ dùng ở fun bên trên thôi ạ

**Review by vietnh-1109 — APPROVED** [review]

File: app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt [hunk 7]
```diff
@@ -2,20 +2,96 @@ package com.sun.efood.ui.home
...
+        randomMeals.add(meal)
+        randomMealAdapter.updateData(randomMeals)
+    }
+
+    override fun showMealCategory(mealCategory: List<MealCategory>) {
+        mealCategoryAdapter.updateData(mealCategory.toMutableList())
+    }
+
+    override fun showMessage(messageRes: Any) {
+        Toast.makeText(context, messageRes.toString(), Toast.LENGTH_SHORT).show()
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/efood/ui/home/HomeFragment.kt]: e nên viết thành extension function để còn dùng lại về sau nhé

File: app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            override fun onSuccess(data: List<MealCategory>) {
+                view.showMealCategory(data)
+                view.hideLoading()
+            }
+
+            override fun onFailure(e: Exception?) {
+                view.showMessage(e?.message.toString())
+                view.hideLoading()
+
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/efood/ui/home/HomePresenter.kt]: remove blank

**Review by thonc-2135 — APPROVED** [review]

**Pull request merged**